### DISCUSSION TITLE
How to make one workflow dependent on the other?

### DISCUSSION THREAD

**Discussion by sayakpaul** [community/community#26907] [Actions]
<p>I am operating on the following two workflows -</p>
<ul>
<li>
<a href="https://github.com/sayakpaul/wine/blob/03db96c0e2d35e78cfee089c7f48290245230009/.github/workflows/commit.yaml" rel="nofollow noopener"><code>commit.yaml</code></a> that operates on <code>push</code> triggers.</li>
<li>
<a href="https://github.com/sayakpaul/wine/blob/03db96c0e2d35e78cfee089c7f48290245230009/.github/workflows/pr_new.yaml" rel="nofollow noopener"><code>pr_new.yaml</code></a> that operates on <code>pull_request</code> triggers.</li>
</ul>
<p>I would want to make <code>pr_new.yaml</code> dependent on <code>commit.yaml</code>. What should I be looking for?</p>
<p>Here’s my repository (along with the desired branch): <a href="https://github.com/sayakpaul/wine/tree/experiment-1/" rel="nofollow noopener">https://github.com/sayakpaul/wine/tree/experiment-1/</a>.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a>,</p>
<p>It’s not supported to set <code>pull_request</code> event depending on <code>push</code> event.</p>
<p>Typically there are two methods to set ‘dependence’:</p>
<ol>
<li>Use two jobs in one workflow, the job2 <code>needs</code> job1.</li>
</ol>
<pre><code class="lang-auto">jobs:
  job1:
  job2:
    needs: job1
</code></pre>
<p>Doc here: <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds" rel="nofollow noopener">https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds</a></p>
<ol start="2">
<li>Use event ‘workflow_run’, it can be used as a trigger when another workflow is requested or completed.</li>
</ol>
<pre><code class="lang-auto">on:
  workflow_run:
    workflows: ["Run Tests"]
    branches: [main]
    types: 
      - completed
      - requested
</code></pre>
<p>Doc here: <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run" rel="nofollow noopener">https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run</a></p>
<p>The limitation is you cannot get the pull request number in both methods( ${{ github.event.number }}). Instead, you can get all pull request of the repo via <a href="https://developer.github.com/v3/pulls/#list-pull-requests" rel="nofollow noopener">rest api</a>, parse the pull request whose head ref is your branch to get the pull request number.</p>
<p>Thanks</p>

**Comment by sayakpaul** [comment]
<p>Thanks, <a class="mention" href="/weide-zhou">@weide-zhou</a>. The problem I am facing is that the <code>pull_request</code> trigger is fetching the backdated values from <code>run.txt</code> which gets updated during the commit. If you see this <a href="https://github.com/sayakpaul/wine/pull/7" rel="nofollow noopener">PR</a> the two run URLs differ from each other:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/f/fbc81a12341a1669e4e41206de23a5d7fb1646c9.png" data-download-href="/uploads/short-url/zVmyND5FU9qkOFVJNuWq3EMIiFP.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/22957388/181100560-258061d9-16a1-4de5-ac7a-51f559c8e1f4.png" alt="image" data-base62-sha1="zVmyND5FU9qkOFVJNuWq3EMIiFP" width="690" height="143" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/f/fbc81a12341a1669e4e41206de23a5d7fb1646c9_2_690x143.png, https://aws1.discourse-cdn.com/github/original/2X/f/fbc81a12341a1669e4e41206de23a5d7fb1646c9.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/f/fbc81a12341a1669e4e41206de23a5d7fb1646c9.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/f/fbc81a12341a1669e4e41206de23a5d7fb1646c9_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">937×195 17.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>The commit message one is the latest while the commented one is the older one.</p>
<p>Any ways to solve this other than making the workflows independent?</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a>,</p>
<p>So the problem is how to make sure the 2nd workflow get the latest run.txt url.<br>
Since it’s not supported to depend pull request event on push event, you can use two jobs as a workaround(not ideal but should work).</p>
<pre><code class="lang-auto">name: test
on: [push]

jobs:
  commit:
    runs-on: ubuntu-latest
    outputs:
      string: ${{ steps.changestring.outputs.string}}
    steps:
      - uses: actions/checkout@v2
      - name: generate new url and change to run.txt
         run: | 
              ....... # Your steps to generate new url and store to run.txt
      - name: set as output
        id: changestring
        run: |
          echo "::set-output name=string::$(&lt;run.txt)"

  PR-Comment:
    runs-on: ubuntu-latest
    needs: [commit]
    steps:
      - name: add PR issue comment
        run: |
          PRnumber=$(curl -s -H "Accept: application/vnd.github.v3+json" https://api.github.com/repos/${{github.repository}}/pulls?head=${{github.repository_owner}}:${{github.ref}} | jq -r '.[] | .number')
          echo "::set-env name=body::${{needs.commit.outputs.string}}"
          echo "::set-env name=number::$(echo $PRnumber)"
      - name: PR Comment
        uses: actions/github-script@v3
        with:
          github-token: ${{secrets.GITHUB_TOKEN}}
          script: |
            github.issues.createComment({
              issue_number: ${{ env.number}},
              owner: context.repo.owner,
              repo: context.repo.repo,
              body: '${{ env.body }}'
            })
</code></pre>
<p>Now you get the PR number and new URL, you can use them to add PR comment.</p>
<p>It works on my side, please check my <a href="https://github.com/weide-zhou/ticket34/runs/1100470183?check_suite_focus=true" rel="nofollow noopener">workflow</a> for your reference.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/77f959f473c9c52b43ea28293128bae6c639fdba.png" data-download-href="/uploads/short-url/h7l3dmOxGR4XMZpHNUN1HssXzLA.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181100563-0f3753ba-1a88-4328-a065-c79b41b65c7e.png" alt="image" data-base62-sha1="h7l3dmOxGR4XMZpHNUN1HssXzLA" width="690" height="278" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/7/77f959f473c9c52b43ea28293128bae6c639fdba_2_690x278.png, https://aws1.discourse-cdn.com/github/optimized/2X/7/77f959f473c9c52b43ea28293128bae6c639fdba_2_1035x417.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/7/77f959f473c9c52b43ea28293128bae6c639fdba.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/77f959f473c9c52b43ea28293128bae6c639fdba_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1088×439 26.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
Note: This works when one pull request from head branch, if you have mutiple PRs, you can get a list of pull request number, and use a loop to add PR comment.
<p>Thanks</p>
<p>Edited: fixed env code, and provide a sample workflow.</p>

**Comment by sayakpaul** [comment]
<p>Thank you very much for the detailed explanation. I tried adapting your workflow (see <a href="https://github.com/sayakpaul/wine/blob/3c7695ec8bd6916f70539a660f99154204ae0135/.github/workflows/commit.yaml" rel="nofollow noopener">here</a>) but the <code>body</code> is not getting set and I am not sure why. Am I missing out on something?</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a>,</p>
<p>I notice the body is empty in your ‘PR-Comment’ job (value of <code>${{needs.commit.outputs.string}}</code>), this is due to you didn’t set it in the dependent job ‘commit’. Please fix it and give a try. :grinning:</p>
<pre><code class="lang-auto">jobs:
  commit:
    runs-on: ubuntu-latest
    outputs:
      string: ${{ steps.changestring.outputs.string}}    # set the job output in commit job then it can be got from PR-Comment job.
</code></pre>
<p>Thanks</p>

**Comment by sayakpaul** [comment]
<p><a class="mention" href="/weide-zhou">@weide-zhou</a> tried that but still no luck. Workflow file <a href="https://github.com/sayakpaul/wine/blob/153741637a4d7eb61305fec96e4752b044c2ee33/.github/workflows/commit.yaml" rel="nofollow noopener">here</a>. Workflow output <a href="https://github.com/sayakpaul/wine/actions/runs/253054515" rel="nofollow noopener">here</a>.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a>,</p>
<p>The <code>set-ouput</code> command in your job <code>commit</code> doesn’t work correctly, it seems the two commands messed.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/70d196def3e2dce9030dc4f0ab2f689e41595891.png" data-download-href="/uploads/short-url/g62we1VQhLEJSvzGDEOvckKJuYp.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181100568-00d6b8f1-41ad-4875-8169-588a8aeb9725.png" alt="image" data-base62-sha1="g62we1VQhLEJSvzGDEOvckKJuYp" width="690" height="106" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/70d196def3e2dce9030dc4f0ab2f689e41595891_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1024×158 8.86 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Please remove <code>cat</code> command, make sure the output value is correctly set. (your error should be due the run.txt format, since if i directly create run.txt on github web, the cat and set-output commands work well).</p>
<pre><code class="lang-auto">      - name: Store the run URL
        id: changestring
        run: |
          echo "::set-output name=string::$(cat run.txt)"
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/9/9a24ae4ebb15f84748e2a1becaa155649e5e555e.png" data-download-href="/uploads/short-url/lZC7n13XW47D9AOZN5CHJdC7R9k.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181100569-041f90b7-d017-4642-ad48-ed6ed9a97492.png" alt="image" data-base62-sha1="lZC7n13XW47D9AOZN5CHJdC7R9k" width="690" height="205" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/9/9a24ae4ebb15f84748e2a1becaa155649e5e555e_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">745×222 16.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
Thanks.

**Comment by sayakpaul** [comment]
<p>I see. I just tried your fix. It works seamlessly now :slight_smile:<br>
Thank you!</p>
